### PR TITLE
[pytorch][build] update nccl to 2.12.12 & aws-ofi-nccl to 1.4.0-aws

### PR DESCRIPTION
**Pull request by harishneit**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [x] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [x] `sagemaker_local_tests = true`

## Label Checklist
- [x] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [x] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [x] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [x] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pytorch/training/docker/1.11/py3/cu113/Dockerfile.gpu [hunk 1]
```diff
@@ -46,11 +46,11 @@ ENV TORCH_CUDA_ARCH_LIST="3.7 5.0 7.0+PTX 8.0"
 ENV TORCH_NVCC_FLAGS="-Xfatbin -compress-all"
 
 ENV CUDNN_VERSION=8.2.0.53
-ENV NCCL_VERSION=2.10.3
+ENV NCCL_VERSION=2.12.12
 ENV HOROVOD_VERSION=0.24.3
 ENV EFA_VERSION=1.15.2
 ENV OMPI_VERSION=4.1.1
-ENV BRANCH_OFI=1.2.0-aws
+ENV BRANCH_OFI=1.4.0-aws
```
Comment (Qingzi-Lan) [inline pytorch/training/docker/1.11/py3/cu113/Dockerfile.gpu:53]: Hi @harishneit, let's change it to `aws` [branch](https://github.com/aws/aws-ofi-nccl/tree/aws), so it always pointing to the latest OFI_NCCL version.

File: pytorch/training/docker/1.11/py3/cu115/Dockerfile.gpu [hunk 2]
```diff
@@ -45,11 +45,11 @@ ENV PATH /opt/conda/bin:$PATH
 ENV TORCH_CUDA_ARCH_LIST="3.7 5.0 7.0+PTX 8.0"
 ENV TORCH_NVCC_FLAGS="-Xfatbin -compress-all"
 
-ENV NCCL_VERSION=2.10.3
+ENV NCCL_VERSION=2.12.12
 ENV HOROVOD_VERSION=0.24.3
 ENV EFA_VERSION=1.15.2
 ENV OMPI_VERSION=4.1.1
-ENV BRANCH_OFI=1.2.0-aws
+ENV BRANCH_OFI=1.4.0-aws
```
Comment (Qingzi-Lan) [inline pytorch/training/docker/1.11/py3/cu115/Dockerfile.gpu:52]: same here

File: pytorch/training/docker/1.11/py3/cu113/Dockerfile.gpu (hunk 1 above)
Comment (harishneit) [reply pytorch/training/docker/1.11/py3/cu113/Dockerfile.gpu:53]: Relying on a proper release tag is safer as we can expect `aws-ofi-nccl` to do some QA before creating the release.  We can't expect the same with a branch.

File: pytorch/training/docker/1.11/py3/cu115/Dockerfile.gpu (hunk 2 above)
Comment (harishneit) [reply pytorch/training/docker/1.11/py3/cu115/Dockerfile.gpu:52]: See previous comment.

File: pytorch/training/docker/1.11/py3/cu113/Dockerfile.gpu (hunk 1 above)
Comment (Qingzi-Lan) [reply pytorch/training/docker/1.11/py3/cu113/Dockerfile.gpu:53]: checked with Rashika from OFI_NCCL. the `aws` branch is a stable branch, and it is safe to use it. DLAMI also uses `aws` branch in its build process. https://github.com/aws/deep-learning-amis/blob/master/scripts/integrations/efa/build_efa_nccl_plugin.py#L41

I think we can follow the same in DLC

Comment (harishneit) [reply pytorch/training/docker/1.11/py3/cu113/Dockerfile.gpu]: Great! If we have that guarantee we could use the branch.

Comment (harishneit) [reply pytorch/training/docker/1.11/py3/cu113/Dockerfile.gpu]: But one issue will be the container builds will not be reproducible, do we want to guarantee that?

Comment (Qingzi-Lan) [reply pytorch/training/docker/1.11/py3/cu113/Dockerfile.gpu]: like other unpinned packages, we can always use the latest stable version of OFI-NCCL. IMHO, the package itself should maintain the backward compatibility, performance and etc to make sure it's reproducible.

Comment (harishneit) [reply pytorch/training/docker/1.11/py3/cu113/Dockerfile.gpu:53]: As discussed offline, let's used the pinned version in this PR to preserve the reproducibility of container build.

**Review by Qingzi-Lan — APPROVED** [review]

**Review by Qingzi-Lan — DISMISSED** [review]

**Comment by Qingzi-Lan** [comment]
please check this pr
https://github.com/aws/deep-learning-containers/pull/2136

**Comment by harishneit** [comment]
@Qingzi-Lan PR 2136 doesn't update NCCL

**Comment by tejaschumbalkar** [comment]
PT1.11 is out of support

**Pull request closed without merging**